### PR TITLE
Deps upgrade

### PR DESCRIPTION
**Pull request by kzemek**
This PR upgrades most deps to their newest versions, preferring Hex deps over git ones.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[5256.1](https://travis-ci.org/esl/MongooseIM/jobs/408082071) / Erlang 19.3 / small_tests / df283cbbc01fe65e75d16c5d982eda6cd6a3e7ce
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5256/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5256/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-25_14.56.51/lib.mongooseim.logs/run.2018-07-25_14.56.52/suite.log.html)

**Comment by mongoose-im** [comment]
[5258.1](https://travis-ci.org/esl/MongooseIM/jobs/408104237) / Erlang 19.3 / small_tests / 8b55815eece21c2d7a152e9bf569036b1e3de0cb
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5258/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5258/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-25_16.08.23/lib.mongooseim.logs/run.2018-07-25_16.08.24/suite.log.html)

**Comment by mongoose-im** [comment]
[5268.1](https://travis-ci.org/esl/MongooseIM/jobs/408495737) / Erlang 19.3 / small_tests / 6969695c02c437026eb23ffa22ea769ebffe77ab
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5268/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5268/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-26_13.17.12/lib.mongooseim.logs/run.2018-07-26_13.17.13/suite.log.html)

**Comment by mongoose-im** [comment]
[5269.1](https://travis-ci.org/esl/MongooseIM/jobs/408525786) / Erlang 19.3 / small_tests / 021c3a1ca87b4cb0ad0cbb0db084f898c55d1989
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5269/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5269/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-26_14.07.30/lib.mongooseim.logs/run.2018-07-26_14.07.30/suite.log.html)

**Comment by mongoose-im** [comment]
[5270.1](https://travis-ci.org/esl/MongooseIM/jobs/408856184) / Erlang 19.3 / small_tests / e57a49810cf16f7f58ecaab961fb943e97800dff
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5270/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5270/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-27_09.07.20/lib.mongooseim.logs/run.2018-07-27_09.07.21/suite.log.html)

**Comment by mongoose-im** [comment]
[5271.1](https://travis-ci.org/esl/MongooseIM/jobs/408875252) / Erlang 19.3 / small_tests / b961273c4cc53c7e860b5e1b57e846f1ad52b78c
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5271/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5271/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-27_10.04.50/lib.mongooseim.logs/run.2018-07-27_10.04.51/suite.log.html)

**Comment by mongoose-im** [comment]
[5272.1](https://travis-ci.org/esl/MongooseIM/jobs/408902232) / Erlang 19.3 / small_tests / 652a064e5576cdbd602bfde2942bfe26c9e8404f
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5272/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5272/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-27_11.32.08/lib.mongooseim.logs/run.2018-07-27_11.32.09/suite.log.html)

**Comment by mongoose-im** [comment]
[5273.1](https://travis-ci.org/esl/MongooseIM/jobs/408920723) / Erlang 19.3 / small_tests / 9ad81e34aa310f581f2172d2cf8f965094b9e667
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5273/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5273/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-27_12.25.18/lib.mongooseim.logs/run.2018-07-27_12.25.19/suite.log.html)

**Comment by mongoose-im** [comment]
[5278.1](https://travis-ci.org/esl/MongooseIM/jobs/408963535) / Erlang 19.3 / small_tests / dcafa2ddb81544222066a58a2d3e45b23c950354
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5278/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5278/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-27_14.42.24/lib.mongooseim.logs/run.2018-07-27_14.42.25/suite.log.html)

**Comment by mongoose-im** [comment]
[5282.1](https://travis-ci.org/esl/MongooseIM/jobs/409022733) / Erlang 19.3 / small_tests / 08cd4210ee809baa24e0c01c0253c7ca26e5571e
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5282/small_tests.19.3)

**Comment by mongoose-im** [comment]
[5285.1](https://travis-ci.org/esl/MongooseIM/jobs/409028702) / Erlang 19.3 / small_tests / 231154d02b06c81721f74ae41c312a790d49fb58
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5285/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5285/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-27_17.07.06/lib.mongooseim.logs/run.2018-07-27_17.07.07/suite.log.html)

**Comment by mongoose-im** [comment]
[5344.9](https://travis-ci.org/esl/MongooseIM/jobs/413036493) / Erlang 21.0 / riak_mnesia / 3d7a3ea9e5c4d225440c783ed14a1e405ac90e16
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5344/riak_mnesia.21.0)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5344/riak_mnesia.21.0/big/ct_run.test%40travis-job-656126bf-57df-4b9d-82b8-a010ae5ea371.2018-08-07_10.13.26/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5344/riak_mnesia.21.0/small/ct_run.mongooseim@localhost.2018-08-07_10.05.39/lib.mongooseim.logs/run.2018-08-07_10.05.40/suite.log.html)
**OK: 1289** / Failed: 0 / **User-skipped: 43** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[5349.1](https://travis-ci.org/esl/MongooseIM/jobs/413053812) / Erlang 19.3 / small_tests / be60539947b600bb7c2f047fa0b072572da6674b
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5349/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5349/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-07_10.51.33/lib.mongooseim.logs/run.2018-08-07_10.51.34/suite.log.html)

**Comment by mongoose-im** [comment]
[5352.1](https://travis-ci.org/esl/MongooseIM/jobs/413094354) / Erlang 19.3 / small_tests / 62b40622b86c2d06857e700dcb1ef7158d908913
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5352/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5352/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-07_12.44.35/lib.mongooseim.logs/run.2018-08-07_12.44.35/suite.log.html)

**Comment by mongoose-im** [comment]
[5353.1](https://travis-ci.org/esl/MongooseIM/jobs/413115649) / Erlang 19.3 / small_tests / 1b91aaae7ff293c525fd15687aad460571419199
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5353/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5353/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-07_13.35.17/lib.mongooseim.logs/run.2018-08-07_13.35.17/suite.log.html)

**Comment by mongoose-im** [comment]
[5354.1](https://travis-ci.org/esl/MongooseIM/jobs/413141060) / Erlang 19.3 / small_tests / 56af0421d78fbe2d0907b33490065510b8ff4171
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5354/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5354/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-07_14.26.04/lib.mongooseim.logs/run.2018-08-07_14.26.05/suite.log.html)

File: src/mongoose_api_json.erl [hunk 1]
```diff
@@ -38,39 +38,49 @@ serialize(Data) ->
...
+    integer_to_binary(Key);
+prepare_key(Key) ->
+    prepare_str(Key).
+
+prepare_str(List) when is_list(List) ->
+    try unicode:characters_to_binary(List) of
+        Bin when is_binary(Bin) -> Bin;
+        _ -> List
+    catch
+        error:badarg -> List
```
Comment (fenek) [inline src/mongoose_api_json.erl]: Should we fall back to original list here? Maybe we should crash if we don't get a valid unicode here?

Comment (kzemek) [reply src/mongoose_api_json.erl]: This is a kludge to try and encode lists that are actually strings - lists that are actually lists should be encoded as JSON arrays instead

**Comment by kzemek** [comment]
@fenek `cpool` wasn't updated because global distrib seems to work better with the buggy logic of version 0.1.0. `worker_pool` was only updated to 3.1.1 because 4.0 requires OTP>=21.

File: src/mongoose_api_json.erl (hunk 1 above)
Comment (kzemek) [reply src/mongoose_api_json.erl]: To rephrase, we don't know if the list is a string or not - so we try to encode it to check.

**Comment by fenek** [comment]
OK, it's understandable in case of `worker_pool`. As for `cpool` - probably we could attempt to rewrite global distrib with `worker_pool` (with some improvements). So I understand that nothing is left?

**Comment by kzemek** [comment]
Yes

File: src/mongoose_api_json.erl (hunk 1 above)
Comment (fenek) [reply src/mongoose_api_json.erl]: Please rename this function then or add a comment because this semantics is not obvious.

**Comment by mongoose-im** [comment]
[5379.1](https://travis-ci.org/esl/MongooseIM/jobs/417253952) / Erlang 19.3 / small_tests / 9819b8bae37deb1a2fb4d557b25ee0ba11c0360e
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2002/5379/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2002/5379/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-17_13.06.10/lib.mongooseim.logs/run.2018-08-17_13.06.11/suite.log.html)

**Pull request merged**